src/nvim/terminal.c

```diff
@@ -152,6 +153,81 @@ static VTermScreenCallbacks vterm_screen_callbacks = {
   .sb_popline  = term_sb_pop,
 };
 
+/// URL decoding (also know as Percent-encoding).
+///
+/// Note this function currently is only used for decoding shell's
+/// OSC 7 escape sequence which we can assume all bytes are valid
+/// UTF-8 bytes. Thus we don't need to deal with invalid UTF-8
+/// encoding bytes like 0xfe, 0xff.
+static size_t url_decode(const char *src, const size_t len, char_u *dst)
+{
+  size_t i = 0, j = 0;
+
+  while (i < len) {
+    if (src[i] == '%' && i + 2 < len) {
+      dst[j] = (char_u)hexhex2nr((char_u *)&src[i + 1]);
+      j++;
+      i += 3;
+    } else {
+      dst[j] = (char_u)src[i];
+      i++;
+      j++;
+    }
+  }
+  dst[j] = '\0';
+  return j;
+}
+
+/// Sync terminal buffer's cwd with shell's pwd with the help of OSC 7.
+///
+/// The OSC 7 sequence has the format of
+/// "\033]7;file://HOSTNAME/CURRENT/DIR\033\\"
+static void sync_shell_dir(const char *command, size_t cmdlen, void *user)
+{
+  size_t offset = strlen("file://");
+  const char *pos = command + offset;
+  char_u *new_dir;
+
+  // remove HOSTNAME to get PWD
+  while (*pos != '/' && offset < cmdlen) {
+    offset += 1;
+    pos += 1;
+  }
+
+  if (offset >= cmdlen) {
+    semsg(_("E1179: Failed to extract PWD from %s, check your shell's config related to OSC 7"),
+          command);
+    return;
+  }
+
+  new_dir = xcalloc(1, cmdlen - offset);
```

Comment:
url_decode adds NUL so xmalloc'd be fine

Reply:
btw i think the actual max len is `cmdlen - offset - 1` to make up for the VTermStringFragment changes (-2, +1 for NUL)